pubtools/_pulp/tasks/push/items/base.py

```diff
@@ -281,11 +298,30 @@ def refresh_after_copy(_):
 
             return f
 
+        # A helper function to re-try copying of failed items
+        #  Since the association phase can take a while and the items can be
+        #  moved to other repos meanwhile, we fetch up-to-date info about the
+        #  items from pulp and re-try the failing cases
+        def fix_failures(_):
+            # Get an up-to-date version of all the copy items.
+            f = cls.items_with_pulp_state_single_batch(pulp_client, copy_items)
+
+            copy_results = []
+            results = list(f.result())
+            if len(results) > 0:
+                # copy any items
+                _, _, copy_results = enqueue_items(results)
+
+            # Return a future with list of all results of copy operations
+            return f_sequence(copy_results)
+
         # This future completes once *all* copies are done successfully.
         # TODO: this still could be improved, as not every item needs every copy
         # before the state could be refreshed.
         all_copies = f_sequence(copy_results)
 
+        all_copies = f_map(all_copies, fix_failures)
```

Comment:
Single retry might work as expected in current scenario. 
Ideally, the failure that occurred once might occur again. It still the same scenario that could be influenced by number of items that failed, association time pulp is takes and the how frequently the chosen source repo updates.  Hence, this may be retired more than once. The retry count should be defined by some environment variable, that's configured in the target settings.
Other than that this looks good,

Reply:
Thanks, changed.